onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py

```diff
@@ -0,0 +1,474 @@
+    batch_size, past_seq_len = 2, 8
+    num_heads, head_size = config.num_attention_heads, int(config.hidden_size / config.num_attention_heads)
+    input_ids = torch.randint(low=0, high=config.vocab_size, size=(batch_size, 1), dtype=torch.int64)
+    attn_mask = torch.randint(low=0, high=2, size=(batch_size, past_seq_len + 1), dtype=torch.int64)
+    # pos_ids is of shape (batch_size, 1)
+    pos_ids = attn_mask.long().cumsum(-1) - 1
+    pos_ids.masked_fill_(attn_mask == 0, 1)
+    pos_ids = pos_ids[:, -1].unsqueeze(-1)
+    past_kv = [
+        (
+            torch.rand(batch_size, num_heads, past_seq_len, head_size),
+            torch.rand(batch_size, num_heads, past_seq_len, head_size),
+        )
+        for _ in range(config.num_hidden_layers)
+    ]
+
+    return (input_ids, attn_mask, pos_ids, past_kv)
+
+
+def get_model_dynamic_axes(input_names: List[str], output_names: List[str]):
+    dynamic_axes = {}
+    for name in input_names + output_names:
+        if name in input_names:
+            # shape is (batch_size, sequence_length)
+            dynamic_axes[name] = {0: "batch_size", 1: "sequence_length"}
+        elif name == "logits":
+            # shape is (batch_size, sequence_length, vocab_size)
+            dynamic_axes[name] = {0: "batch_size", 1: "sequence_length"}
+        elif "present" in name:
+            # shape is (batch_size, num_heads, past_sequence_length + 1, head_size)
+            dynamic_axes[name] = {0: "batch_size", 2: "past_sequence_length + 1"}
+        else:
+            raise Exception("Unknown input or output name found")
+    return dynamic_axes
+
+
+def get_model_with_past_kv_dynamic_axes(input_names: List[str], output_names: List[str]):
+    dynamic_axes = {}
+    for name in input_names + output_names:
+        if name in {"input_ids", "position_ids"}:
+            # shape is (batch_size, 1)
+            dynamic_axes[name] = {0: "batch_size"}
+        elif name == "attention_mask":
+            # shape is (batch_size, past_sequence_length + 1)
+            dynamic_axes[name] = {0: "batch_size", 1: "past_sequence_length + 1"}
+        elif "past" in name:
+            # shape is (batch_size, num_heads, past_sequence_length, head_size)
+            dynamic_axes[name] = {0: "batch_size", 2: "past_sequence_length"}
+        elif name == "logits":
+            # shape is (batch_size, 1, vocab_size)
+            dynamic_axes[name] = {0: "batch_size"}
+        elif "present" in name:
+            # shape is (batch_size, num_heads, past_sequence_length + 1, head_size)
+            dynamic_axes[name] = {0: "batch_size", 2: "past_sequence_length + 1"}
+        else:
+            raise Exception("Unknown input or output name found")
+    return dynamic_axes
+
+
+def save_onnx_model(onnx_model: onnx.ModelProto, output_path: str, data_path: str):
+    onnx.save(
+        onnx_model,
+        output_path,
+        save_as_external_data=True,
+        all_tensors_to_one_file=True,
```

Comment:
Is it better to save one tensor per file?

Reply:
I set the flag as true by default because it is easier to have one data file vs lots of data files when moving the model and its external data between locations.